### ISSUE TITLE
Regex Parser - consider adding an option to cache regex and result

### ISSUE THREAD

**Issue by djaglowski** [open-telemetry/opentelemetry-log-collection#41]
In some cases, a regex does not need to be evaluated very often. For example, when parsing a filename, as seen [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/2266#issuecomment-788132157).

A major performance gain may be possible in such cases, if a mechanism can be designed that would allow the regex parser to cache strings and their results.

The user interface to this could be as simple as adding a `cacheable: true` option the the parser config.

Things that need to be considered:
- How does the cache behave if enabled for high cardinality operations?
- A cache should be available only to an instance of the operator.
- How is the result applied to the entry? (There is likely an obvious answer, but this needs validation.)
- Does the proposed `cacheable: true` make sense? Perhaps a `caching` block is necessary, which would contain further options such as `max_size`, etc.

**Comment by jsirianni** [comment]
I am working on this, as today.

**Comment by jsirianni** [comment]
I have a working concept right now

```yaml
filelog:
  include:
  - /var/log/containers/*.log
  start_at: end
  operators:
  - type: regex_parser
    # NOTE: The trailing dollar sign is escaped with a second dollar sign
    regex: '^(?P<pod_name>[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*)_(?P<namespace>[^_]+)_(?P<container_name>.+)-(?P<container_id>[a-z0-9]{64})\.log$$'
    parse_from: $$attributes["file.name"]
    cache:
      type: memory
      size: 100
```

Regex parser takes a `cache` block which is actually part of the parser oporator struct. The cache is an interface with the following methods:
```
// Takes a key and returns the parsed value from a previous regex operation
Get(key string) (interface{}, bool)

// Takes a key and parsed result, adds it to the cache
Add(key string, data interface{})

// Useful for testing, allows the test code to dump the cache and inspect it. 
Copy() map[string]interface{}

// Useful for testing, allows the test code to dump the configured max size
MaxSize() uint16
```

The cache type looks like this
```
type Memory struct {
	cache map[string]item
	keys chan string
	mutex sync.RWMutex
}
```

Entries are stored using a string key and an item (interface{}) value. Interface was choosen because that is the datatype passed to and returned by all parsers.

A channel is used to keep track of the queue, allowing easy removal of the oldest key when room needs to be made.

All reads / writes lock with the appropriate read / write lock.

Each regex operator can have its own cache, and it would be encouraged that users only cache operations that make sense, like parsing a k8s container log file name, where the result is **always** the same until the pod / container goes away and is replaced by a new one (and a new log file with a new name).

The cache defaults to a size of 100 but is configurable. When the cache is full, the oldest entry is removed in favor of the newest entry. 

Some quick Go benchmarks. The first iteration will store the result in the cache, future iterations are strictly read operations unless the cache is full (at which point Go routines will need to wait for a write lock to finish adding a new entry):
```
// go test -bench=. ./operator/parser/regex -benchtime=2s -benchmem

// Single regex, single go routine. Basically comparing the regex operation to a map read. Results are not surprising here. A map read will always be significantly faster when it is not blocked by a mutex.
BenchmarkParseNoCacheOneFile-32              	  347572	      6874 ns/op	     664 B/op	       8 allocs/op
BenchmarkParseWithMemoryCacheOneFile-32      	285175263	         8.826 ns/op	       0 B/op	       0 allocs/op

// Parsing 100 unique inputs with 100 Go routines, the cache is not full
BenchmarkParseNoCache-32                     	   15564	    160684 ns/op	   96151 B/op	    1005 allocs/op
BenchmarkParseWithMemoryCache-32             	   45367	     52966 ns/op	    6435 B/op	     201 allocs/op

// Parsing 100 unique inputs with 100 Go routines, cache size of 99 (always replacing the oldest with the newest)
BenchmarkParseWithMemoryCacheFullByOne-32    	   38998	     64277 ns/op	    8895 B/op	     212 allocs/op

// Same as previous but cache size is 90
BenchmarkParseWithMemoryCacheFullBy10-32     	   21488	    112591 ns/op	   19323 B/op	     309 allocs/op

// Same as previous but cache size is 50
BenchmarkParseWithMemoryCacheFullBy50-32     	   10951	    268281 ns/op	   55922 B/op	     650 allocs/op
```

Performance is very good, compared to regex. Performance drops if the cache is constantly full, always replacing old entries with new entries. On Kubernetes, the cache size should be set to the number of pods per node (which is a configuration option on kubernetes) + number of side cars (it is possible to have 100 pods with 2 containers each, 200 log files.)

I suspect that in the real world, only some containers are logging at a high rate while others are not, meaning not all files will be competing for cache space in the event that the cache is full. 


Happy to discuss next week
- cache code: https://github.com/jsirianni/opentelemetry-log-collection/tree/feature/parser-cache/operator/cache
- regex cache code: https://github.com/jsirianni/opentelemetry-log-collection/blob/feature/parser-cache/operator/parser/regex/regex.go#L106

**Comment by djaglowski** [comment]
Great progress on this. It certainly seems this could increase performance in some cases.

A few initial thoughts:
- It's not immediately clear to me that this functionality would benefit other parsers besides the regex parser. From what I have seen, other parsers tend to be more performant and have fewer use cases that involve parsing a small set of values repeatedly (e.g. a set of file names that don't change often). I think a stronger case should be made if we're going to apply it across the board.
- It would be interesting to see a few more data points that show even higher levels of churn. It appears that exceeding the cache size has a very negative effect on performance - it would be good to understand how extreme that becomes at higher ratios. I wonder if at some level of churn the cache should just be skipped?

**Comment by jsirianni** [comment]
I did a couple things:
- Moved all cache logic to the regex package
- Renamed all cache interface / types / globals so they are not exported (except for CacheConfig)
- Added a benchmark where the cache has a max size of 10 and we try to add 100 items
- Added a benchmark where the cache has a max size of 1 and we try to add 100 items
- Added a mechanism for throttling the Add() method when the cache is full and has had eviction recently

TLDR: The throttling mechanism gives the slow benchmarks a chance to compete with regex. Without throttling, they are very slow (half speed or less).

The rate limiting implementation may be somewhat naive, I am not sure. I'll need to give it some more thought.

You can find it here: https://github.com/jsirianni/opentelemetry-log-collection/tree/feature/parser-cache-performance/operator/parser/regex
